proposals/kroxylicious-operator-api-alternative.md

```diff
@@ -0,0 +1,654 @@
+
+## CRDs
+
+* Proxy CR - an instance of the Kroxylicious
+* ProxyIngress CR - Defines a way to access a Proxy
+* VirtualCluster CR - a virtual cluster
+* Filter CR - a filter definition
+
+![image](https://github.com/user-attachments/assets/8767da6b-3fc4-4fbc-99bc-64e93c06bc70)
+
+https://excalidraw.com/#json=CVAZMl-MuN1QP2cIW3hcA,cEWi5VjUg1mK2N1lvU-_xQ
+
+
+## Personas / Responsibilies
+
+* Infrastructure admin - responsible for the proxy tier and how proxies are exposed to the network.
+* Developer - responsible for the virtualcluster and filter
+
+## API
+
+### Proxy
+
+The Proxy CR is the responsibilty of the the Infrastructure admin.
+
+Declares an instance of the proxy and defines the ingress mechanisms available to it.
+
+```yaml
+apiVersion: proxy.kroxylicious.io/v1
+kind: Proxy
+metadata:
+  name: myproxy
+spec:
+
+   # Optional - controls the Proxy infrastructure.
+   infrastructure:
+      # Influences the annotations and labels on all resources created by the operator.
+      annotations:
+         x: y
+      labels:
+         x: y
+      proxy:
+        # Controls specific to operand deployment
+        replicas: 2
+        resources:
+          requests:
+            memory: "64Mi"
+            cpu: "250m"
+#  # Optional: default ordered list of filters to be used by a virtualcluster that doesn't apply it own.  (this doesn't fit well with the Infra role - drop it?).
+#  defaultFilterRefs:
+#  - group: filter.kroxylicious.io
+#    kind: Filter
+#    name: encryption```
+
+status:
+   # status of the proxy tbd
+   conditions:
+   - status: x
+     type: y
+     reason: r
+   # per listener status 
+   ingresses:
+   - name: myclusterip
+     condidtions:
+     - ...
+```
+
+### ProxyIngress
+
+The ProxyIngress CR is the responsibilty of the the Infrastructure admin.
+
+It declares a named ingress mechanism - in other words, a way for traffic to get to a proxy.  ProxyIngress knows about four types:
+
+* clusterIP - for on-cluster traffic realised using Kubernetes ClusterIP Service. Supports TCP or TLS.
+* loadBalancer - for off-cluster realised using a Kubernetes LoadBalancer Service.  Supports TLS only.
+* openShiftRoute - for off-cluster realised using OpenShift Routes (OpenShift specific)
+* gateway - for off-cluster realised using Gateway API.  Supports `TLSRoutes` with listener type of `passthrough`.
+
+```yaml
+apiVersion: proxy.kroxylicious.io/v1
+kind: ProxyIngress
+metadata:
+  name: myclusterip
+spec:
+  # reference to the proxy that will use this ingress.
+  proxyRef:
+    kind: Proxy  # if present must be Gateway, otherwise defaulted
+    group: proxy.kroxylicious.io # if present must be proxy.kroxylicious, otherwise defaulted=
+    name: myproxy  # name of proxy
+
+  # oneOf: clusterIP, loadBalancer, openShiftRoute, gateway
+  clusterIP:
+    protocol: TCP|TLS (ClusterIP supports both)
+    port: # optional, defaults to 9082
+  loadBalancer:
+    bootstrapAddressPattern: $(vitrtualClusterName).kafka.com 
```

Comment:
So `virtualClusterName` is an Operator only concept that lets one Ingress be used for multiple VirtualClusters?

Reply:
`$(vitrtualClusterName)` would be understood by Kroxylicious, rather than the Operator.  Both  `$(vitrtualClusterName)`  and `$(nodeId)` would be passed down as literals into the proxy config file.  `$(vitrtualClusterName)` would be substituted for the virtual cluster's name.  $(nodeId) is as-is.

btw.  I did wonder about pushing bootstrapAddressPattern and brokerAddressPattern into the VirtualCluster.  That would do away with the need introduce the new pattern.